### PR TITLE
[WIP] Add application grouping to taskbar

### PR DESCRIPTION
**Pull request by kubanecxxx**
Fixes lxde/lxqt#132.

### CODE REVIEW

**Comment by kubanecxxx** [comment]
First of all I wish you happy new year guys.
The plugin should be ready now I fixed those issues.

File: plugin-taskbar/lxqttaskbutton.h [hunk 1]
```diff
@@ -89,25 +98,40 @@ public slots:
...
     void mouseReleaseEvent(QMouseEvent *event);
     void mouseMoveEvent(QMouseEvent *event);
-    void contextMenuEvent(QContextMenuEvent *event);
+    virtual void contextMenuEvent(QContextMenuEvent *event);
     void paintEvent(QPaintEvent *);
 
+    void setWindowId(WId wid) {mWindow = wid;}
+    virtual void arbitraryMimeData(QMimeData * mimedata);
+    virtual void draggingTimerTimeout() {activateWithDraggable();}
+    virtual QString acceptMimeData() const {return taskButtonMimeDataFormat();}
```
Comment (jleclanche) [inline plugin-taskbar/lxqttaskbutton.h]: What are you doing here? This just implicitly returns `QString("lxqt/lxqttaskbutton")`. Is there a reason these two methods can't go away in favour of a #define?

File: plugin-taskbar/lxqtmasterpopup.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    l->setSpacing(0);
+    l->setMargin(0);
+
+    connect(mCloseTimer,SIGNAL(timeout()),this,SLOT(close()));
+    mCloseTimer->setSingleShot(true);
+    mCloseTimer->setInterval(400);
+}
+
+
+void LxQtMasterPopup::moveAnimated(const QPoint &newpos)
```
Comment (jleclanche) [inline plugin-taskbar/lxqtmasterpopup.cpp]: This and resizeAnimated() are both unnecessary imo, they are just used in the one place and there's no reason they would be used anywhere else. Thoughts @luis-pereira ?

File: plugin-taskbar/lxqttaskbutton.h (hunk 1 above)
Comment (kubanecxxx) [inline plugin-taskbar/lxqttaskbutton.h]: It is a virtual method and its called in base class LxQtTaskButton::dragEnterEvent - it is overrided if you drag group or button.

**Comment by paulolieuthier** [comment]
I get a crash everytime I try to reorder the buttons with grouping disabled. Index out of range somewhere.
```
(gdb) bt full
#0  0x00007ffff5a38a97 in raise () from /usr/lib/libc.so.6
No symbol table info available.
#1  0x00007ffff5a39e6a in abort () from /usr/lib/libc.so.6
No symbol table info available.
#2  0x00007ffff666325f in QMessageLogger::fatal(char const*, ...) const ()
   from /usr/lib/libQt5Core.so.5
No symbol table info available.
#3  0x00007ffff665e5d1 in qt_assert_x(char const*, char const*, char const*, int) ()
   from /usr/lib/libQt5Core.so.5
No symbol table info available.
#4  0x00007ffff7bb11b9 in QList<QLayoutItem*>::move (this=0x786180, from=-1, to=0)
    at /usr/include/qt/QtCore/qlist.h:639
No locals.
#5  0x00007ffff7bb01ec in LxQt::GridLayout::moveItem (this=0x777eb0, from=-1, to=0)
    at /home/paulo/Workspace/lxqt/lxde-qt/liblxqt/lxqtgridlayout.cpp:334
        d = 0x786180
#6  0x00007fffdb9a04c0 in LxQtTaskBar::groupDroppedSlot (this=0x77ae10, point=..., 
    event=0x7fffffffc860)
    at /home/paulo/Workspace/lxqt/lxqt-panel/plugin-taskbar/lxqttaskbar.cpp:209
        droppedIndex = -1
        __PRETTY_FUNCTION__ = "void LxQtTaskBar::groupDroppedSlot(const QPoint&, QDropEvent*)"
        newPos = 0
        p = 164
        size = 3
#7  0x00007fffdb9b1db2 in LxQtTaskBar::qt_static_metacall (_o=0x77ae10, 
    _c=QMetaObject::InvokeMetaMethod, _id=5, _a=0x7fffffffc4e0)
    at /home/paulo/Workspace/lxqt/lxqt-panel/build/plugin-taskbar/moc_lxqttaskbar.cpp:95
        _t = 0x77ae10
#8  0x00007ffff688ff69 in QMetaObject::activate(QObject*, int, int, void**) ()
   from /usr/lib/libQt5Core.so.5
No symbol table info available.
#9  0x00007fffdb9b2329 in LxQtTaskButton::dropped (this=0x78afd0, _t1=..., 
    _t2=0x7fffffffc860)
    at /home/paulo/Workspace/lxqt/lxqt-panel/build/plugin-taskbar/moc_lxqttaskbutton.cpp:233
        _a = {0x0, 0x7fffffffc520, 0x7fffffffc4c8}
#10 0x00007fffdb9a4ea1 in LxQtTaskButton::dropEvent (this=0x78afd0, event=0x7fffffffc860)
    at /home/paulo/Workspace/lxqt/lxqt-panel/plugin-taskbar/lxqttaskbutton.cpp:185
No locals.
#11 0x00007ffff718be68 in QWidget::event(QEvent*) () from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#12 0x00007ffff730d689 in QToolButton::event(QEvent*) () from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#13 0x00007ffff7148d5c in QApplicationPrivate::notify_helper(QObject*, QEvent*) ()
   from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#14 0x00007ffff714e688 in QApplication::notify(QObject*, QEvent*) ()
---Type <return> to continue, or q <return> to quit---
   from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#15 0x00007ffff68606cb in QCoreApplication::notifyInternal(QObject*, QEvent*) ()
   from /usr/lib/libQt5Core.so.5
No symbol table info available.
#16 0x00007ffff71a55bc in ?? () from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#17 0x00007ffff71a60bb in ?? () from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#18 0x00007ffff7148d5c in QApplicationPrivate::notify_helper(QObject*, QEvent*) ()
   from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#19 0x00007ffff714e340 in QApplication::notify(QObject*, QEvent*) ()
   from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#20 0x00007ffff68606cb in QCoreApplication::notifyInternal(QObject*, QEvent*) ()
   from /usr/lib/libQt5Core.so.5
No symbol table info available.
#21 0x00007ffff6b911ac in QGuiApplicationPrivate::processDrop(QWindow*, QMimeData const*, QPoint const&, QFlags<Qt::DropAction>) () from /usr/lib/libQt5Gui.so.5
No symbol table info available.
#22 0x00007ffff6b76fc9 in QWindowSystemInterface::handleDrop(QWindow*, QMimeData const*, QPoint const&, QFlags<Qt::DropAction>) () from /usr/lib/libQt5Gui.so.5
No symbol table info available.
#23 0x00007fffed143e91 in ?? () from /usr/lib/qt/plugins/platforms/libqxcb.so
No symbol table info available.
#24 0x00007fffed145dce in ?? () from /usr/lib/qt/plugins/platforms/libqxcb.so
No symbol table info available.
#25 0x00007ffff6b846ad in QBasicDrag::eventFilter(QObject*, QEvent*) ()
   from /usr/lib/libQt5Gui.so.5
No symbol table info available.
#26 0x00007ffff6860383 in QCoreApplicationPrivate::sendThroughApplicationEventFilters(QObject*, QEvent*) () from /usr/lib/libQt5Core.so.5
No symbol table info available.
#27 0x00007ffff7148ce8 in QApplicationPrivate::notify_helper(QObject*, QEvent*) ()
   from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#28 0x00007ffff714e340 in QApplication::notify(QObject*, QEvent*) ()
   from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#29 0x00007ffff68606cb in QCoreApplication::notifyInternal(QObject*, QEvent*) ()
   from /usr/lib/libQt5Core.so.5
No symbol table info available.
#30 0x00007ffff6b8fdab in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) () from /usr/lib/libQt5Gui.so.5
No symbol table info available.
#31 0x00007ffff6b91785 in QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*) () from /usr/lib/libQt5Gui.so.5
---Type <return> to continue, or q <return> to quit---
No symbol table info available.
#32 0x00007ffff6b76f1f in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Gui.so.5
No symbol table info available.
#33 0x00007fffed164160 in ?? () from /usr/lib/qt/plugins/platforms/libqxcb.so
No symbol table info available.
#34 0x00007ffff471da0d in g_main_context_dispatch () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#35 0x00007ffff471dcf8 in ?? () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#36 0x00007ffff471ddac in g_main_context_iteration () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#37 0x00007ffff68b7ca7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Core.so.5
No symbol table info available.
#38 0x00007ffff685e162 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) ()
   from /usr/lib/libQt5Core.so.5
No symbol table info available.
#39 0x00007ffff6b84a52 in QBasicDrag::drag(QDrag*) () from /usr/lib/libQt5Gui.so.5
No symbol table info available.
#40 0x00007ffff6b9c811 in QDragManager::drag(QDrag*) () from /usr/lib/libQt5Gui.so.5
No symbol table info available.
#41 0x00007ffff6b9bb87 in QDrag::exec(QFlags<Qt::DropAction>, Qt::DropAction) ()
   from /usr/lib/libQt5Gui.so.5
---Type <return> to continue, or q <return> to quit---
No symbol table info available.
#42 0x00007fffdb9a52f7 in LxQtTaskButton::mouseMoveEvent (this=0x7a9580, 
    event=0x7fffffffda20)
    at /home/paulo/Workspace/lxqt/lxqt-panel/plugin-taskbar/lxqttaskbutton.cpp:261
        mime = 0x775010
        drag = 0x96f640
        pixmap = <incomplete type>
#43 0x00007ffff718be68 in QWidget::event(QEvent*) () from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#44 0x00007ffff730d689 in QToolButton::event(QEvent*) () from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#45 0x00007ffff7148d5c in QApplicationPrivate::notify_helper(QObject*, QEvent*) ()
   from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#46 0x00007ffff714e976 in QApplication::notify(QObject*, QEvent*) ()
   from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#47 0x00007ffff68606cb in QCoreApplication::notifyInternal(QObject*, QEvent*) ()
   from /usr/lib/libQt5Core.so.5
No symbol table info available.
#48 0x00007ffff714d54a in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool) () from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#49 0x00007ffff71a3e30 in ?? () from /usr/lib/libQt5Widgets.so.5
---Type <return> to continue, or q <return> to quit---
No symbol table info available.
#50 0x00007ffff71a5fa3 in ?? () from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#51 0x00007ffff7148d5c in QApplicationPrivate::notify_helper(QObject*, QEvent*) ()
   from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#52 0x00007ffff714e340 in QApplication::notify(QObject*, QEvent*) ()
   from /usr/lib/libQt5Widgets.so.5
No symbol table info available.
#53 0x00007ffff68606cb in QCoreApplication::notifyInternal(QObject*, QEvent*) ()
   from /usr/lib/libQt5Core.so.5
No symbol table info available.
#54 0x00007ffff6b8fdab in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) () from /usr/lib/libQt5Gui.so.5
No symbol table info available.
#55 0x00007ffff6b91785 in QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*) () from /usr/lib/libQt5Gui.so.5
No symbol table info available.
#56 0x00007ffff6b76f1f in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Gui.so.5
No symbol table info available.
#57 0x00007fffed164160 in ?? () from /usr/lib/qt/plugins/platforms/libqxcb.so
No symbol table info available.
#58 0x00007ffff471da0d in g_main_context_dispatch () from /usr/lib/libglib-2.0.so.0
---Type <return> to continue, or q <return> to quit---
No symbol table info available.
#59 0x00007ffff471dcf8 in ?? () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#60 0x00007ffff471ddac in g_main_context_iteration () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#61 0x00007ffff68b7ca7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Core.so.5
No symbol table info available.
#62 0x00007ffff685e162 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) ()
   from /usr/lib/libQt5Core.so.5
No symbol table info available.
#63 0x00007ffff6865b3c in QCoreApplication::exec() () from /usr/lib/libQt5Core.so.5
No symbol table info available.
#64 0x0000000000411a57 in main (argc=1, argv=0x7fffffffe438)
    at /home/paulo/Workspace/lxqt/lxqt-panel/panel/main.cpp:111
        configFile = {static null = {<No data fields>}, 
          d = 0x7ffff6937020 <QArrayData::shared_null>}
        app = 0x64bf80
        term = {__sigaction_handler = {sa_handler = 0x4115e6 <termSignalHandler(int)>, 
            sa_sigaction = 0x4115e6 <termSignalHandler(int)>}, sa_mask = {__val = {
              0 <repeats 16 times>}}, sa_flags = 272818081, sa_restorer = 0xff00000000}
        res = false
```

**Comment by paulolieuthier** [comment]
@kubanecxxx fixed, thanks. I think there's only two comments I still like to make:
* The animation is a little weird when using the panel on the bottom, as the animated group appears from top to bottom, I think it would be ideal to slide from bottom to top (when panel on botom) or something like that.
* Second thing, and much more important: I think the way the code is organized and the logic is implemented is conceptually wrong. LXQtTaskGroup inherits from LXQtTaskButton, but holds LXQtTaskButtons inside itself. When creating the taskbar, the LXQtTaskbar class creates LXQtTaskGroups to have only one window (if grouping is disabled) or many windows (if grouping disabled).

Although it works the way it is now (and we could probably shiped it), IMO we should try to make everything look as much correct and elegant as we can. I think there should be an abstract class named LXQtAbstractButton, that would be inherited from both LXQtTaskGroup and LXQtTaskButton. The task groups would contain buttons inside, but the taskbar would contain groups **OR** buttons, according to grouping settings. It would probably be nice as well to have an utils class, to reduce the size of the classes to the minimum necessary.

@kubanecxxx, what is your opinion on that? @jleclanche, @luis-pereira, @PCMan?

**Comment by kubanecxxx** [comment]
I agree with you it definitely consumes more memory and it needs refactoring. 
I am not sure with the LXQtAbstractButton because I need whole LXQtTaskButton functionality in LXQtTaskGroup with single window.

I like your idea about groups or buttons in taskbar according to the group setting.

What should be the next step in this plugin? Should I first try to refactor it or are you going to merge it?

**Comment by paulolieuthier** [comment]
> I am not sure with the LXQtAbstractButton because I need whole LXQtTaskButton functionality in LXQtTaskGroup with single window.

I see. What I meant was to add all that common functionality to a super class. That way the taskbar could also accept **both** groups and buttons. Maybe that's the best way to go.

> What should be the next step in this plugin? Should I first try to refactor it or are you going to merge it?

Well, let's start the refactoring. The first step is to work on the super class and make every class as simple as possible. I'll land your PR on a separate branch and create a new PR from that new branch, so we can both commit to it (and others as well).

**Comment by paulolieuthier** [comment]
Actually, we woudn't be able to commit directly. Let's keep it that way, and I will send PRs to your branch. If you accept them and push, they will get into this PR here.

**Comment by paulolieuthier** [comment]
I've been still thinking of the design of our implementation. Maybe enhancing the taskbutton class to be able to handle multiple windows if needed could be the best idea. I'm now wondering how Plasma does it. We could make use of advices here.

**Comment by jleclanche** [comment]
I'm closing this PR, I merged it into the "grouping" branch and work needs to happen there now.

**Pull request closed without merging**